### ISSUE TITLE
Testing DOSBox-X translation using gettext

### ISSUE THREAD

**Issue by aybe** [joncampbell123/dosbox-x#2693]
A quick test is available here: https://github.com/joncampbell123/dosbox-x/tree/translation

Not sure what's going on with the escape codes though:

<img width="361" alt="dosbox-x_2021-07-21_06-27-13" src="https://user-images.githubusercontent.com/1537591/126431481-e011d39b-9a35-44b0-aeb2-16886dfe3ef9.png">

**Most importantly, the way it is used is weird, .PO files are supposed to be like this:**

```
msgid "source text"
msgid "translated text"
...
```

But they are like this:

```
msgid "AUTOEXEC_CONFIGFILE_HELP"
msgstr ""
"Lines in this section will be run at startup.\n"
"You can put your MOUNT lines here.\n"
"\n"
```

So in the end, they don't follow the rule, forcing one to peek onto the original en-US to find out the original text, pretty boring.

Even more boring: re-write tons of code to conform to that fact; not sure if it's a really good idea.

Now, if say you'd use something like VS Code with synchronized scroll bars, with on the left, the original, on the right the translation, you effectively completely remove the need of gettext and poedit and can stay on the original system which already works.

The only thing that might be of interest in this branch then is [what I did here](https://github.com/joncampbell123/dosbox-x/commit/e46cf4d58f70979c277f995a8ac12849e761f8a0), replace those ◙ U+25D9 by new lines so one doesn't need to use incredibly boring horizontal scrollbars in addition to vertical ones.

TL;DR; the good ol' system is quite effective, upgrading might not be that worth in the end

**Comment by aybe** [comment]
Take 2 😀

As I was translating manually to FR, I started using DeepL which really has amazing translation.

I then realized this could be automated and then review process would be a walk in the park.

There you go, a program that translates .LNG files:

https://github.com/joncampbell123/dosbox-x/tree/translation/contrib/translations/dosbox-translation-deepl

<img width="674" alt="VsDebugConsole_2021-07-21_08-24-56" src="https://user-images.githubusercontent.com/1537591/126442127-c997ba7e-946e-4939-a646-c459f18b2fd0.png">

<img width="674" alt="cmd_2021-07-21_08-32-09" src="https://user-images.githubusercontent.com/1537591/126442231-d67aa2ad-9eaf-4793-b468-6d734e4f9b41.png">

You just need to register for a free API key and you're good: https://www.deepl.com/docs-api/accessing-the-api/

An example of translation:

https://github.com/joncampbell123/dosbox-x/blob/translation/contrib/translations/deepl%20translations/fr_FR.lng

```
:AUTOEXEC_CONFIGFILE_HELP
Les lignes de cette section seront exécutées au démarrage.
Vous pouvez placer vos lignes MOUNT ici.
.

:CONFIGFILE_INTRO
# Ceci est le fichier de configuration pour DOSBox-X %s. (Veuillez utiliser la dernière version de DOSBox-X)
# Les lignes commençant par un # sont des lignes de commentaires et sont ignorées par DOSBox-X.
# Ils sont utilisés pour documenter (brièvement) l'effet de chaque option.
# Pour écrire TOUTES les options, utilisez la commande 'config -all' avec les options -wc ou -writeconf.
.
```
It's amazingly good, just needs some manual review 😎

As for the fixing, I'd go VSCode like this:

<img width="1920" alt="Code_2021-07-21_08-34-39" src="https://user-images.githubusercontent.com/1537591/126442465-3450108b-62b5-452b-8fe9-f6bced9e5e0f.png">

Done using right-click on Explorer side bar:

- Select for Compare
- Compare with Selected

But as you can see FR is problematic, if you want to be clear you have to be verbose and end up being much longer than US:

<img width="361" alt="dosbox-x_2021-07-21_08-38-59" src="https://user-images.githubusercontent.com/1537591/126442991-93e03aba-d6ef-4100-944c-1fd1c2b1b252.png">

OTOH you can go terse but then it sounds dull... translation is a job in itself, not easy.

All in all, I think pre-translating the bulk will undoubtedly help because it relieves the user from painful typing and he can focus on improvements.

**Comment by Wengier** [comment]
@aybe Thanks a lot for comparing between the two approaches and the finding! Based on your find I think keeping using language files is likely a better idea for the purpose of DOSBox-X. But I think the line ending issue (◙ U+25D9) does need to be fixed as you said. So I fixed it both in the code and in language files. As a result, trying to generate language files will properly output the newline character instead of the boring character ◙ as you said. Thanks!

**Comment by aybe** [comment]
Nice, when I get some free time, I'll manually edit the FR translation, it's on my desktop. 😉

**Comment by Wengier** [comment]
@aybe After the fix I have been looking at your French translation, and very glad to see you had almost done it. Wonderful work, and great tool you had introduced. This tool can definitely help others too. In the language file itself, it looks like some lines in the welcome message has overflown the screen, so they need to be shortened. For example, for the following line:

```
Tapez HELP pour voir la liste des commandes du shell, INTRO pour une brève introduction.
```

It is apparently too long for a screen line. It is possible to shorten such lines, as long as they are clearly understandable by readers?

Once again, thanks a lot for the translation! Will add to the installer when done. Great work!

**Comment by aybe** [comment]
As for getting translation done soon, it's a no, loud and clear.

I've been trying to translate for the last 2 hours and it's extremely long and error-prone, I've only been able to fix intro screen and some configuration tool stuff.

I've only done that and barely scratched the surface:

https://github.com/joncampbell123/dosbox-x/commit/8782143b30cfae2fa2bf3f5eba97fce808049a7a

Also, the order in the .LNG file does not help but not sure if it's fixable anyhow.

You progress really slowly because you have to test every time until it's right. And I ran the program 100s of times to simply check the results!

Furthemore, those ANSI codes are sometimes poorly written and can be edited to gain these precious spaces to be able to fit long sentences; that adds up to the process.

Also, some of these should be addressed to make the process easier:

#2698 
#2699
#2700
#2701 
#2702 
#2703

This pattern seems to be a good approach, you can quickly make changes and compare, maybe this could be put on some translation guide:

![2021-07-21_22-06-04](https://user-images.githubusercontent.com/1537591/126553039-589badbf-06d4-4b61-b57a-aeec3e5e4e05.png)

As you can see, lots of compromises to make it sound natural, in the mean time some errors have slipped in (ANSI colors).

**Comment by aybe** [comment]
@Wengier 

https://github.com/joncampbell123/dosbox-x/commit/ffb1606d39a7e84ea6646d4789c31c143495ddb6

That'll have to wait, I thought I could tackle it today but I barely made any progress...

**It's just a massive job, way bigger than one might think:**

- you have to find what to translate and how it looks in program
- then you have to find that line in the LNG
  - the order of entries in the LNG is Bogosort, very boring 🤬
- providing you haven't already forgotten the text (because they are many variations)
  - you can translate but then you have to double, triple check
  - if it's between ANSI your mileage may vary (not an expert in calculating lengths of text intertwined with ANSI, u know...🤪)
- VSCode is surprisingly slow
- I probably opened that menu more times than one may do in a lifetime 😁
- I've only translated the one-liners, just imagine for those big chunks mixed with ANSI
- Sometimes you have to search online in dictionnaries else it's really ambigous, that adds up a little moe
- etc, etc, etc

When it comes to software, FR totally sucks... personally I have all keyboard and operating system in EN-US !

In short, it's very long, spent already 4 hours on it and I'm far from completion, but I've committed it anyway so it's not lost.

For the next victim who may want to continue the work:

- welcome screen has been translated
- the first half of menus have been translated
- some of configuration tool has been
- everything else hasn't been

**Comment by aybe** [comment]
@Wengier 

Wanted to add, when you look at the commit, you may read it in like 30 seconds but believe me, it's a +2 hours job.

Think of it like it's code, imagine handling a file of 3719 lines of C code, which is how long the .LNG is, some may argue LOC isn't a metric but it tells quite a bit; it isn't trivial at all!

Two things that may help:

- edit the mini-program I wrote that pre-translates from DeepL
  - sort by identifiers before outputting the .LNG, a 30secs job thanks to LINQ
- do the same for existing translation,
  - should be a 30mins job, the reading logic is already in that mini-program

Doing so would certainly ease the translation because I suppose there is some kind of logic in the identifiers of items to be translated so, having them in order is certainly interesting to try.

**Comment by aybe** [comment]
@Wengier 

I've added:

- the tools to the branch
- wrote the sorter app.
  - logic has been split, the deepl translator should be fine since I've further improved the parsing, didn't test coz' it;s really late
- added existing translations sorted, next to others, should be fine but didn't overwrite them just in case

https://github.com/joncampbell123/dosbox-x/tree/develop/contrib/translations

**Comment by aybe** [comment]
@Wengier
@joncampbell123 

Check this: 

https://github.com/joncampbell123/dosbox-x/blob/develop/contrib/translations/docs/README.translation.MD

https://github.com/joncampbell123/dosbox-x/tree/develop/contrib/translations/docs

Anyone reading it should be able to do translations, the process seems right to me but should be tested and possibly improved.

Also, there are two things still left to do in the TODO section, see how these should be addressed.

**Comment by aybe** [comment]
Here are builds of the program, apparently the new `Portable` format makes them able to run on any platform:

[dosbox-translation-tools.zip](https://github.com/joncampbell123/dosbox-x/files/6871632/dosbox-translation-tools.zip)

For Windows it works but for the others I don't know because I don't have them.

They will need .NET 5 runtime:

https://docs.microsoft.com/en-us/dotnet/core/install/windows?tabs=net50
https://docs.microsoft.com/en-us/dotnet/core/install/linux
https://docs.microsoft.com/en-us/dotnet/core/install/macos

I tried to build self-contained ones but they each do 60Mb, that would have made a whopping 360Mb zip, so abandoned the idea.

**Comment by aybe** [comment]
@Wengier 

https://github.com/joncampbell123/dosbox-x/blob/develop/contrib/translations/fr/fr_FR.lng
https://github.com/joncampbell123/dosbox-x/blob/develop/contrib/translations/fr/fr_FR%20translation%20notes.txt

Two thirds done, the last third is an order of magnitude longer, don't expect it in a reasonable time frame!

Then there will be the fixing phase, normalization phase and finally improvement phase...

Honestly? FR totally sucks for software... When literature tries to describes technic but fails in doing so by being so verbose.

Even though manual editing clearly improves the situation, it's by a small margin, look at this overflow fest:

![dosbox-x_2021-07-24_07-23-09](https://user-images.githubusercontent.com/1537591/126858319-a80273a2-6536-4095-9bd2-547c654afc70.png)

This is the kind of stuff remaining to do, one will clearly have to make compromises to ensure it fits:

![dosbox-x_2021-07-24_07-29-44](https://user-images.githubusercontent.com/1537591/126858473-b4d3d4f8-d6a0-47c1-a0d5-b9321a75562a.png)

I now understand why so many abbreviations are used for this language in software!

# 🤣

**Comment by FredBezies** [comment]
Indeed, there are overflow everywhere. I'll wait until to finish the first draft, then I will edit the biggest typo. Fixing overflow will be the longer part.

Anyway, I won't be able to use your tools. I stopped using MS-Windows back in 2006.

**Comment by Wengier** [comment]
@aybe Thanks a lot for the translation guide and translation fixes. They are greatly helpful indeed.

For the overflow issue, I think the biggest restriction is still on the screen width - the standard 80-lines can not be overcome. So I guess one has to make compromises in such cases.

**Comment by FredBezies** [comment]
I find the fix for help output.

In the current translation, you have an empty line a dot and another empty line. Which is wrong.

```
:SHELL_CMD_CHDIR_HELP
Affiche ou modifie le répertoire actuel.

.

:SHELL_CMD_CHDIR_HELP_LONG
```

But in the english file, there is an empty line, a dot and then a new entry. Not an useless empty line which breaks everything.

```
:SHELL_CMD_CHDIR_HELP
Displays or changes the current directory.

.
:SHELL_CMD_CHDIR_HELP_LONG
```

When I follow the empty line and the dot like in the english file, it works better. As a proof of concept, I modified every single command invoked with help.

I think attached screenshot speaks for itself :)

![Capture d’écran de 2021-07-24 14-04-30](https://user-images.githubusercontent.com/8571419/126867878-0f87e80f-56c8-4766-b136-790bdb1ae0c2.png)

**Comment by FredBezies** [comment]
This afternoon I took 2 hours to fix all the badly placed empty lines, and fixed some translations here and there. I made a small video to show what I have now in stock on my HDD.

Besides some lines that need to be modified in length, it is a previous good preview of what Dosbox-X can look like in french. I used Dosbox video recorder to show everything.

If you want my translation file, just tell me. It will save a lot of hours of work.

https://user-images.githubusercontent.com/8571419/126871658-52a7a9a0-6cf1-47ca-9224-ccab6638cb32.mp4

**Comment by Wengier** [comment]
@FredBezies Thanks for the updated language file. It is certainly an improvement over the previous one. Meanwhile, you may want to look at the help messages for DOS shell commands, e.g. ``DIR /?``. They are apparently a few overflows that may be adjusted to make them look better.

**Comment by Wengier** [comment]
@FredBezies I have already updated the language file to use the updated one. Is it possible to cleanup the messages a bit more by the next version for inclusion in the installer?

**Comment by Wengier** [comment]
@aybe I have added a link to the translation guide you made to the README page. Hope it will help others too.

**Comment by FredBezies** [comment]
> @FredBezies Thanks for the updated language file. It is certainly an improvement over the previous one. Meanwhile, you may want to look at the mapper editor and help messages for DOS shell commands, e.g. `DIR /?`. They are apparently a few overflows that may be adjusted to make them look better.

I will look at help messages. For mapper part, I cannot do anything before next tuesday.

Will post an updated version for help part asap :)

**Comment by FredBezies** [comment]
> @FredBezies I have already updated the language file to use the updated one. Is it possible to cleanup the messages a bit more by the next version for inclusion in the installer?

Another patch with a more cleaner help output and fixes for PC-98 intro is available. Feel free to merge it. See https://github.com/joncampbell123/dosbox-x/pull/2717

Only missing part? Mapper editor, the more annoying one. Will work on it next tuesday.

**Comment by Wengier** [comment]
@aybe The translation guide appears to be no longer available from the page, so the link to it in README is now broken:

https://github.com/joncampbell123/dosbox-x/blob/develop/contrib/translations/docs/README.translation.MD

Can you post the guide page again so that it will be again accessible? Thanks.

**Comment by aybe** [comment]
Today's a great day:

- nothing of that in history, they were in the repo so how is that even possible?
- don't have the files because as I constantly run short of disk space I have to make choices
- realized that stupid IDrive hasn't been backing up for weeks because of *unexpected error*

**Comment by Wengier** [comment]
@aybe Wow. I think the original page was in the develop branch, which was probably later force-synchronized with the master branch and got lost. I wonder if there is a way to recover pages like this in general? The following page about Git may be useful:

https://git-scm.com/book/en/v2/Git-Internals-Maintenance-and-Data-Recovery
